tools/validate_spec.py

```diff
@@ -79,81 +78,125 @@ def cli(argv: List[str]) -> Union[int, str]:
+            args.directory_valid,
+            args.json_output,
+            args.abort_on_error,
+        )
     except ValidationError as e:
         return f"{e}"
 
     return 0
 
 
-def validation_main(args: argparse.Namespace) -> None:
-    # pylint: disable=too-many-locals
-    def get_message_to_validate(message_path: Path, is_valid: bool) -> "OriginalMessage":
-        if not message_path.is_file():
-            raise ValidationError(f"{message_path} is not a regular file")
-        message_bytes = message_path.read_bytes()
-        message = OriginalMessage(message_bytes, is_valid, message_path.name)
-        return message
-
-    root_message_id: str = args.message_identifier
-    path_spec: Path = args.specification
-    path_invalid: Optional[Path] = args.directory_invalid
-    path_valid: Optional[Path] = args.directory_valid
-    full_output_path: Optional[Path] = args.json_output
-    abort_on_error: bool = args.abort_on_error
-    no_verification: bool = args.no_verification
-
-    path_generators: Dict[Path, Iterator[Path]] = {}
-    for path in [path_valid, path_invalid]:
-        if path is not None:
-            path_generators[path] = path.glob("*")
+def validate(
+    message_value: MessageValue,
+    directory_invalid: Optional[Path],
+    directory_valid: Optional[Path],
+    json_output: Optional[Path],
+    abort_on_error: bool,
+) -> int:
+
+    with OutputWriter(json_output) as output_writer:
+        for directory_path, is_valid_directory in [
+            (directory_valid, True),
+            (directory_invalid, False),
+        ]:
+            directory = directory_path.glob("*") if directory_path is not None else []
+            for path in directory:
+                validation_result = __validate_message(path, is_valid_directory, message_value)
+                output_writer.write_result(validation_result)
+                if not validation_result.validation_success and abort_on_error:
+                    raise ValidationError(f"aborted: message {path} was classified incorrectly")
 
+    return 0
+
+
+def __validate_message(
+    message_path: Path, valid_original_message: bool, message_value: MessageValue
+) -> "ValidationResult":
+    if not message_path.is_file():
+        raise ValidationError(f"{message_path} is not a regular file")
+    original_message = message_path.read_bytes()
+    parser_error: Optional[str] = None
+    parser_result: MessageValue = message_value.clone()
     try:
-        identifier = ID(root_message_id)
+        parser_result.parse(original_message)
+        valid_parser_result = (
+            parser_result.bytestring == original_message and parser_result.valid_message
+        )
+        if not valid_parser_result:
+            parser_error = "Invalid message"
     except RecordFluxError as e:
-        raise ValidationError(f'invalid identifier "{root_message_id}" : {e}') from e
+        parser_error = str(e)
+        valid_parser_result = False
```

Comment:
No, in the `identifier.py` module, a `RecordFluxError` is raised for invalid identifiers.

Reply:
That is a correct statement, but why should such an exception be raised here? There is no instantiation of `ID` here. IIRC `PyRFLXError` was explicitly introduced for the validation tool to be able to only catch parser errors, but now `PyRFLXError` is not used in the validation tool anymore.

Reply:
`ID` is instantiated in line 81 and the exception is catched  in line 82.
> PyRFLXError is not used in the validation tool anymore

That is a correct observation, but PyRFLXError (instead of RecordFLuxError) should be catched in line 147. I fixed that in c1958db.